Implement physics constructor for OpticalPhysics

drbenmorgan opened the pull request:
Celeritas intends to implement optical photon offload from Geant4 via custom `G4VProcess`es for photon-generating processed (Cerenkov, Scintillation). Within Geant4, this means they will need to be added to the relevant particles in application physics lists both to enable the offload and for optical data construction/availablity (as with the EM processes).

Following the pattern for used for EM physics in `CelerEMStandardPhysics`, the changes here add a preliminary concrete `G4VPhysicsConstructor` to handle particle and process construction for Optical processes. Though Celeritas will not generate and track G4OpticalPhotons, this initial implement largely follows the core `G4OpticalPhysics` constructor just to get something in place with `G4OpticalPhoton`. 

Comments are used, though it's fairly obvious, to note where the custom Celeritas processes will go. Adding this constructor in `CelerEmPhysicsList` and/or other lists and `celer-g4` can be done, but wanted to submit a WIP just to check this is on the right base.

drbenmorgan committed:
Implement minimal  physics constructor for OpticalPhysics

Celeritas will implement optical photon offload via custom processes
(Cerenkov, Scintillation), so these will need to be added to the
relevant particles so they are used and optical data loaded on the
Celeritas side.

Follow pattern for EM physics, adding a concrete G4VPhysicsConstructor
to handle particle and process construction for Optical processes.
Though Celeritas will not generate and track G4OpticalPhotons, implement
as per G4OpticalPhysics constructor for initial testing and use. Show
points where the Celeritas processes will go later.

drbenmorgan committed:
Only use G4OpWLS2 with Geant4 10.7 and newer

This process was only introduced in Geant4 10.7, see:

https://geant4.org/download/release-notes/notes-v10.7.0.html

drbenmorgan committed:
Only use G4OpticalParameters with Geant4 >= 10.7

The G4OpticalParameters class was only introduced in Geant4 10.7.
Before this, process activation and configuration is handled in
G4OpticalPhysics directly.

Refactor check on process activation into an anonymous free function
that uses G4OpticalParameters if available and simply returns true
otherwise.

drbenmorgan committed:
Mark argument as maybe_unused for older Geant4 versions

drbenmorgan commented:
Though adding this to `CelerEmPhysicsList` and `CelerFTFPBert` could proceed at this point, it might be worth thinking about how best to handle the optical physics parameters. One gotcha the last few commits show is that `G4OpticalParameters` (analogous to `G4EmParameters`) is only present from Geant4 10.7. Before this, the parameters are handled locally in `G4OpticalPhysics`, which `CelerOpticalPhysics` attempts to emulate.

I think all of the optical parameters can be handled/transfered in the same way as the EM ones are in the `CelerEmStandardPhysics` constructor:

https://github.com/celeritas-project/celeritas/blob/cc2d1df1b3d609c00c12561e2dec60d179e7c012/src/celeritas/ext/detail/CelerEmStandardPhysics.cc#L98-L107

i.e. obtain them from the input `struct` and:

1. For Geant4 < 10.7, hold this locally and configure processes locally (see [how Geant4 did it](https://gitlab.cern.ch/geant4/geant4/-/blob/geant4-10.6-release/source/physics_lists/constructors/electromagnetic/src/G4OpticalPhysics.cc?ref_type=heads#L207-217))
2. Otherwise, call `G4OpticalParameters` setters as appropriate.

Some of this may of course not be needed later if we don't need to set these processes up to get data etc for Celeritas from them.

However, I don't think we can/should add the optical parameters as direct members of `GeantPhysicsOptions` as that's really focussed on EM options. It feels like we should maybe have a separate `GeantOpticalPhysicsOptions` struct that's a member of `GeantPhysicsOptions`? Longer term, could also consider a larger refactor so we'd have something like:

```c++
struct GeantPhysicsOptions
{
    GeantEmPhysicsOptions electromagnetic;
    GeantOpticalPhysicsOptions optical;
    // GeantHadronicPhysicsOptions hadronic;
};
```

but could just start with the optical parameter struct as a member?

sethrj commented:
@drbenmorgan I think your plan for a separate struct (and later creating another one for EM parameters) is perfect. Thanks for this!

drbenmorgan committed:
Add struct for configuring optical physics

Geant4's G4OpticalPhysics component provides a variety of options
to enable and configure the supported optical photon generation
and physics processes. So that Celeritas can configure the Geant4
processes/data and then access them consistently, we want to
mirror this configuration in setting up CelerOpticalPhysics.

Implement GeantOpticalPhysicsOptions struct to hold these options,
following the pattern established for the EM options. Map activation
and process parameters from G4OpticalParameters, available from Geant4
10.7 and newer.

Update CelerOpticalPhysics construction to use an instance of this
struct. Support use with Geant4 < 10.7 with initial workarounds for
process activation and configuration.

More aggressive refactoring will be done once changes in optical
processes between 10.5 and 10.7 are better established.

drbenmorgan committed:
Provide JSON (de)serialization for GeantOpticalPhysicsOptions

drbenmorgan committed:
Correct use of LaTeX in docstring

drbenmorgan committed:
Fix compilation errors with Geant4 < 10.7

drbenmorgan commented:
Struct in place, but as you'll see, the implementation of `CelerOpticalPhysics` is now a mess due to the changes in how optical physics was configured between Geant4 10.5-10.7. I'll have a think how to improve that.

Next is to add `GeantOpticalPhysicsOptions` as a member of `GeantPhysicsOptions`, but ran into some test failures as naturally the generated JSON changes. I did think about having it as a `std::optional` as we probably don't always want optical physics on. However, need to play about with that as I don't think `nlohmann::json` supports this directly (and not sure on Celeritas policy on use of `std::optional`). Advice/ideas welcome!

sethrj commented:
@drbenmorgan To make life easier for now, let's just stick to the newer versions, and we can throw a "not implemented" for the older versions if the experiments still need it by the time we've got sufficient optical physics implemented 😅

drbenmorgan committed:
Add optical options to main GeantPhysicsOptions

Provide GeantOpticalPhysicsOptions in main options structure through
a std::optional to allow clients to enable/disable it as required.
Set as std::nullopt by default to retain existing behaviour/setup.

Update JSON serialization to account for new data member. Provide
minimal from/to_json fully specialized on the held optional type
for simplicity.

Update test that checks physics option JSON to include new data
member.

drbenmorgan committed:
Add GeantOpticalPhysicsOptions to SWIG interface

drbenmorgan committed:
Add CelerOpticalPhysics to physics lists if enabled

Construct/add Celeritas optical physics constructor to overall
physics lists if options for it are supplied.

drbenmorgan commented:
Modulo fixing any errors in the CI, I think this is close to ready. I guess the last step is unit tests for when the optical physics are enabled (they are off by default), but am not 100% sure where to add these. [`GeantImporterTest.cc`](test/celeritas/ext/GeantImporterTest.cc) and the `LarSphere` test would appear to be appropriate, but I then get failures on `G4Cerenkov` and `G4Scintillation` not being exportable by `GeantImporter.cc`. I'm not sure about the higher level `celeb-sim/celer-g4` yet.

sethrj commented on src/celeritas/ext/GeantOpticalPhysicsOptions.hh:
```diff
@@ -0,0 +1,135 @@
+    bool scint_track_info{false};
+    //!@}
+
+    //!@{
+    //! \name Boundary physics options
+    //! Invoke Geant4 SD at post step point if photon deposits energy
+    bool invoke_sd{false};
+    //!@}
+
```
I think this would belong on the higher level setup rather than the physics options...

drbenmorgan replied:
That makes sense, though perhaps we leave for now as a reminder (I can comment out) until there's a mechanism to do this in Celeritas? I'd guess it'll eventually go in `SDSetupOptions` or similar.

drbenmorgan replied:
O.k., latest commit implements the `operator bool` retaining the previous behaviour, so

- default construction enables all Optical processes
- The static function is "deactivated" rather than "all_physics"

So opposite to what you suggested @sethrj, but can easily flip this. It kind of felt more natural for default construction to enable everything, but I don't have a strong preference.

drbenmorgan commented:
Having played a bit with the tests as above, one slight interface smell (depending on point of view) with the use of `std::optional` is that setting options through the top level `GeantPhysicsOptions` feels a bit messy:

```c++
    GeantPhysicsOptions opts; // default, no optical options
    opts.optical_options.cerenkov_radiation = on; // whoops, not the correct interface, and is nullopt anyway

    // so have to either do
    GeantOpticalOptions optical; // gets you the defaults
    optical.cerenkov_radiation = false;
    opts.optical_options = optical;

    // or
    opts.optical_options = GeantOpticalOptions();
    (*opts.optical_options).cerenkov_radiation = false; // or .value()...
```

So I'm leaning towards @sethrj's suggestion to define an `explicit operator bool() const` for `GeantOpticalOptions` which would return `false` if none of the processes are activated. JSON to/from can work just the same (I think), but need to decide what default construction behaviour is, and from that the need for a "`build_default`" or "`deactivate`" member function to provide a simple switch.

sethrj commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,277 @@
+    else if (process_name == "OpRayleigh")
+        return options_.rayleigh_scattering;
+    else if (process_name == "OpWLS")
+        return options_.wavelength_shifting != WLSTimeProfileSelection::none;
```
This  makes me think we might want to move toward changing both EM and optical physics to use a `std::set` of process names rather than an increasing number of bools... makes sense as-is for now though!

sethrj reviewed (commented):
Thanks @drbenmorgan , "optional optical options" does sound a little silly 😅 even though that's what it actually is.

I tend to agree with you that a standalone `operator bool` is the way to go. A couple thoughts:
- We should probably default the optical physics to `off` 
- We can add a `static OpticalPhysicsOptions all_physics()` so that users could assign `optical = ...`

stognini commented on src/celeritas/ext/GeantOpticalPhysicsOptions.hh:
```diff
@@ -0,0 +1,135 @@
+struct GeantOpticalPhysicsOptions
+{
+    //!@{
+    //! \name Optical photon physics
```
Is this a placeholder or was it duplicated by mistake?

drbenmorgan replied:
Yes, a duplication be mistake, so removed!

stognini commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,277 @@
+    scint->SetScintillationTrackInfo(options_.scint_track_info);
+    // These two are not in 10.7 and newer, but defaults should be sufficient
+    // for now
+    // scint->SetScintillationYieldFactor(fYieldFactor);
```
Wait, so we're keeping these commented for now as a reminder to set them up later on? Is the scintillation yield factor here the same as, on newer G4 versions, the scintillation yield that is set via the material properties table?

drbenmorgan replied:
Possibly, I think @sethrj suggested that we focus on 10.7 and newer and revisit here if and only if experiments positively require support for the earlier versions. I'd have to look through the code to check on the yield being the same.

drbenmorgan committed:
Remove extraneous Doxygen comment

drbenmorgan committed:
Use operator bool instead of optional for optical options

Holding optical options as a std::optional in GeantPhysicsOptions
whilst convenient, showed a few usage issues:

- Slightly less clear interface to construct.
- Need to deref/use value() to access/change submembers.

Refactor GeantOpticalOptions with boolean conversion operator that
only returns true if at least one process is activated. Provide
a static member function to return a fully deactivated instance to
retain default construction as fully enabled. Update JSON de/serialization
to output null for a deactivated object, and construct a deactivated
instance on reading null.

Update GeantOptions and Celeritas physics lists to hold/use a bare
instance of GeantOpticalOptions instead of the previous std::optional.

drbenmorgan committed:
Merge remote-tracking branch 'upstream/develop' into celeropticalphysics

sethrj commented on src/celeritas/ext/GeantOpticalPhysicsOptions.hh:
```diff
@@ -0,0 +1,156 @@
+    int cerenkov_max_photons{100};
+    //! Maximum percentage change in particle \f$\beta\f$  before limiting step
+    double cerenkov_max_beta_change{10.0};
+    //!@}
```
We could avoid the scintillation/cerenkov bools and the prefixes here by doing another "optional" struct... but that might be too complicated (or we would end up taking it to the logical conclusion of a set of physics option classes in a hierarchy of their own 🤪)

drbenmorgan replied:
I think that's reasonable as it'd only be for the Cerenkov, Scintillation and Boundary processes. Happy to go ahead and do this if that's the prefered way.

drbenmorgan replied:
O.k. went ahead and did this as it was a pretty simple copy-paste exercise!

sethrj replied:
Excellent, thanks Ben!

sethrj commented on src/celeritas/ext/GeantPhysicsOptions.hh:
```diff
@@ -136,6 +138,10 @@ struct GeantPhysicsOptions
     bool verbose{false};
+
+    //! Optical physics options
+    GeantOpticalPhysicsOptions optical_options{
```
Maybe optical? So we could end up with `options.optical.cerenkov.max_photons`?

sethrj commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,277 @@
+    // TODO: Celeritas will eventually implement these directly (no
+    // G4OpticalPhotons) so how to set up on "Celeritas-side"
+    auto absorption = std::make_unique<G4OpAbsorption>();
+    if (process_is_active("OpAbsorption"))
```
Does geant4 error if there's a typo in these names or just return "false"? It would be safer to have an enum plus string mapping since these strings are also repeated in `process_is_active`. We have something similar in `celeritas/io/ImportProcess.hh`...

Maybe for now we could make string constants local to this file just to avoid the repetition?

stognini replied:
It should return "NoProcess" (associated with enum `G4OpticalProcessIndex::kNoProcess`).
I think the enum string mapping makes sense too.

drbenmorgan replied:
[c4ba704](https://github.com/celeritas-project/celeritas/pull/1348/commits/c4ba704312bd749c8d5b4c0f8fbbea6b9b96a9be) should implement a _rough_ version of this. Everything's internal to `CelerOpticalPhysics.cc` for now, with only the minimal enum-enum/string mapping implemented.

sethrj reviewed (commented):
Looks good @drbenmorgan , this is a somewhat tedious yet tricky and important task! I think the only substantial change to make would be to split off process-dependent options into another level of `struct`.

stognini commented on src/celeritas/ext/detail/CelerEmPhysicsList.cc:
```diff
@@ -32,6 +33,14 @@ CelerEmPhysicsList::CelerEmPhysicsList(Options const& options)
     auto em_standard = std::make_unique<CelerEmStandardPhysics>(options);
     RegisterPhysics(em_standard.release());
+
+    // Celeritas-supported Optical Physics
```
Nitpick:

```suggestion
    if (options.optical_options)
    {
        // Celeritas-supported Optical Physics
```

amandalund commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,277 @@
+    if (process_is_active("OpWLS2"))
+    {
+        process_manager->AddDiscreteProcess(wls2.release());
+        // I need to check how this differs from G4OpWLS...
```
It looks identical except that it uses separate optical properties... so I guess it lets you have two different wavelength shifting processes in the same material?

stognini replied:
I did not realize WLS2 was akin to the scintillation components. That makes sense and we might consider generalizing it like we did with the scintillation components. 

PS: As an e.g., the NOvA detectors use a mineral oil with a dopant that scintillates and at least one other that produces a wavelength shifting. The scintillation/shifted photons get captured by an optical fiber that is also a wavelength-shifting fiber, so technically at least 2 WLS processes happen between scintillation and capture by the fiber that leads to a sensor.

drbenmorgan committed:
Remove _options substring for clearer naming/use

drbenmorgan committed:
Localize comment to code it applies to

drbenmorgan committed:
Implement optical process enum to avoid hardcoded strings

Per review comments, we'd like to avoid direct use of Geant4's internal
string names for optical processes. This is for code robustness (spelling)
and avoiding repitions.

Provide a private OpticalProcessType enumeration over the optical process
types together with helper functions to translate to Geant4 internal
indices and names. Update ConstructProcess to use these functions for
checking process activation.

drbenmorgan committed:
Fix compilation error for Geant4 < 10.7

drbenmorgan committed:
Have structs-per-process for optical physics options

Review suggested organising optical process options into one struct
per process for those having settings beyond simple enablement.
Each struct would hold a flag to indicate that process being enabled,
plus the additional configuration parameters.

Implement this design for Cerenkov, Scintillation, and Boundary
optical processes. Update internal use and JSON serialization to
use new interface.

drbenmorgan committed:
Explicitly convert options structs to bool

sethrj reviewed (approved):
I think this is great, @drbenmorgan , and a lot of painstaking work! Much appreciated.

amandalund reviewed (approved):
Nice work @drbenmorgan, thanks for this addition!

sethrj committed:
Merge branch 'develop' into celeropticalphysics

stognini commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,337 @@
+namespace detail
+{
+namespace
+{
```
```suggestion
{
//---------------------------------------------------------------------------//
/*!
 * Optical physics process type.
 *
 * See Geant4's \c G4OpticalProcessIndex in G4OpticalParameters.hh for the equivalent enum.
 */
```

stognini commented on src/celeritas/ext/GeantOpticalPhysicsOptions.hh:
```diff
@@ -0,0 +1,211 @@
+}
+
+//---------------------------------------------------------------------------//
+//! Optical Bounrary process options
```
```suggestion
//! Optical Boundary process options
```

stognini commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,337 @@
+        case OpticalProcessType::wavelength_shifting:
+            return options.wavelength_shifting != WLSTimeProfileSelection::none;
+        case OpticalProcessType::wavelength_shifting_2:
+            // Not supported pre 10.7
```
I know we're not worrying about any version < 10.7, but should we throw a `CELER_ASSERT_UNREACHABLE();` here?

stognini commented on src/celeritas/ext/detail/CelerOpticalPhysics.cc:
```diff
@@ -0,0 +1,337 @@
+        process_manager->AddDiscreteProcess(wls2.release());
+        // I need to check how this differs from G4OpWLS...
+        CELER_LOG(debug)
+            << "Loaded Optical wavelength shifting V2 with G4OpWLS2 "
```
Since WLS and WLS2 are identical, maybe

```suggestion
            << "Loaded second optical wavelength shifting with G4OpWLS2 process "
```

drbenmorgan commented:
@stognini I'll implement your suggestions in a fresh PR!

sethrj commented:
Sorry I pulled the trigger too soon!

stognini commented:
No worries. Everything was looking great, so was just having a final look in case I spotted any minor detail after yesterday's commits. Thanks a lot @drbenmorgan !
